### PR TITLE
Add beacon chain related new type

### LINKED ISSUE #1383: Add beacon chain / eth2.0 related NewType
### What is wrong?
https://github.com/ethereum/py-evm/pull/1373#discussion_r224107745
To make the types more explict!

### How can it be fixed
1. Time unit - `Second` - and also can be used for PoW chain?
2. For beacon chain, some other `NewType` candidates: `Slot`, `ShardID`. And [with the latest spec](https://github.com/ethereum/eth2.0-specs/blob/cbd254a3090771e79ccc8eb8e21c7b32dba93c45/specs/beacon-chain.md#constants): `ValidatorStatus`, `SpecialRecordType`, `ValidatorSetDeltaFlag`.
3. **EDIT:** Move to [ethereum/eth-typing/](https://github.com/ethereum/eth-typing/)

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

This is an attempt to fix #1383

### How was it fixed?

- Add `NewType`s in `eth/beacon/typing.py`

### Discussion

I added some additional new types:

1. Differentiate signatures in `bytes` and `Tuple[int, int]` form.
```python
BLSSignature = NewType('BLSSignature', bytes)  
BLSSignatureAggregated = NewType('BLSSignatureAggregated', Tuple[int, int]) # Tuple[uint384, uint384]
```
2. Add `Timestamp` and `Second` types.
3. Add a `ValidatorIndex` type.

#### Also, some issues need to resolve

- Are 5th slot and 5 slots considered same `SlotNumber` type?
- Do we need `uint64` and `uint384` types to differentiate different integers? The purpose is to check the intention of the variable rather than to check integer length or signed/unsigned.

#### Suggested types

- CommitteeIndex

### Cute Animal Picture

image

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
After discussion with @hwwhww and @mhchia. Since the nature that all types defined in this PR are only used for the time being, we can just focus on the short-term gain. We conclude that

- `BLSSignature`: Though the distinction between `bytes` and `Tuple[int, int]` only happens in py_ecc. , defining two types for both cases is beneficial in py-evm now. Rename them to `BLSSignatureBytes` and `BLSSignatureIntegers`.
- `uint64` etc. are not urgent recently, and a better decision could be made when https://github.com/ethereum/eth2.0-specs/issues/308 resolves.
- `CommitteeIndex` is good to be added.

For the last question, `Are 5th slot and 5 slots considered same SlotNumber type?`. Should we use a type `ShardCount` for variable `SHARD_COUNT`, or should it share the same type with `ShardNumber`. We would like to request the wisdom from @cburgdorf.

File: eth/beacon/types/exits.py [hunk 1]
```diff
@@ -24,9 +28,9 @@ class Exit(rlp.Serializable):
     ]
 
     def __init__(self,
-                 slot: int,
+                 slot: SlotNumber,
                  validator_index: int,
```
Comment (hwwhww) [inline eth/beacon/types/exits.py]:
```suggestion
                 validator_index: ValidatorIndex,
```

File: eth/beacon/types/shard_committees.py [hunk 2]
```diff
@@ -27,7 +30,7 @@ class ShardCommittee(rlp.Serializable):
     ]
 
     def __init__(self,
-                 shard: int,
+                 shard: ShardNumber,
                  committee: Sequence[int],
```
Comment (hwwhww) [inline eth/beacon/types/shard_committees.py]:
```suggestion
                 committee: Sequence[ValidatorIndex],
```

File: eth/beacon/types/shard_reassignment_records.py [hunk 3]
```diff
@@ -21,8 +25,8 @@ class ShardReassignmentRecord(rlp.Serializable):
 
     def __init__(self,
                  validator_index: int,
```
Comment (hwwhww) [inline eth/beacon/types/shard_reassignment_records.py]:
```suggestion
                 validator_index: ValidatorIndex,
```

File: eth/beacon/types/states.py [hunk 4]
```diff
@@ -76,27 +83,27 @@ class BeaconState(rlp.Serializable):
...
+            justification_bitfield: Bitfield,
+            finalized_slot: SlotNumber,
             processed_pow_receipt_root: Hash32,
             validator_registry: Sequence[ValidatorRecord]=(),
-            validator_balances: Sequence[int]=(),
+            validator_balances: Sequence[Gwei]=(),
             latest_randao_mixes: Sequence[Hash32]=(),
             latest_vdf_outputs: Sequence[Hash32]=(),
             shard_committees_at_slots: Sequence[Sequence[ShardCommittee]]=(),
             persistent_committees: Sequence[Sequence[int]]=(),
```
Comment (hwwhww) [inline eth/beacon/types/states.py]:
```suggestion
            persistent_committees: Sequence[Sequence[ValidatorIndex]]=(),
```

File: eth/beacon/types/validator_registry_delta_block.py [hunk 5]
```diff
@@ -28,7 +32,7 @@ class ValidatorRegistryDeltaBlock(rlp.Serializable):
     def __init__(self,
                  latest_registry_delta_root: Hash32,
                  validator_index: int,
```
Comment (hwwhww) [inline eth/beacon/types/validator_registry_delta_block.py]:
```suggestion
                 validator_index: ValidatorIndex,
```

File: eth/beacon/types/validator_registry_delta_block.py [hunk 6]
```diff
@@ -28,7 +32,7 @@ class ValidatorRegistryDeltaBlock(rlp.Serializable):
...
-                 pubkey: int,
+                 pubkey: BLSPubkey,
                  flag: int) -> None:
```
Comment (hwwhww) [inline eth/beacon/types/validator_registry_delta_block.py]:
```suggestion
                 flag: ValidatorRegistryDeltaFlag) -> None:
```

File: eth/beacon/helpers.py [hunk 7]
```diff
@@ -370,8 +383,8 @@ def get_effective_balance(validator_balances: Sequence[int], index: int, max_dep
 
 
 def get_new_validator_registry_delta_chain_tip(current_validator_registry_delta_chain_tip: Hash32,
-                                               validator_index: int,
-                                               pubkey: int,
+                                               validator_index: ValidatorIndex,
+                                               pubkey: BLSPubkey,
                                                flag: int) -> Hash32:
```
Comment (hwwhww) [inline eth/beacon/helpers.py]:
```suggestion
                                               flag: ValidatorRegistryDeltaFlag) -> Hash32:
```

Comment (hwwhww) [reply eth/beacon/helpers.py]: FYI: I'm removing this function in other PR.

**Comment by ChihChengLiang** [comment]
Thank you @hwwhww  for the review ❤️

File: eth/beacon/types/states.py [hunk 8]
```diff
@@ -76,27 +83,27 @@ class BeaconState(rlp.Serializable):
...
+            validator_registry_latest_change_slot: SlotNumber,
             validator_registry_exit_count: int,
             validator_registry_delta_chain_tip: Hash32,
-            previous_justified_slot: int,
-            justified_slot: int,
-            justification_bitfield: int,
-            finalized_slot: int,
+            previous_justified_slot: SlotNumber,
+            justified_slot: SlotNumber,
+            justification_bitfield: Bitfield,
```
Comment (hwwhww) [inline eth/beacon/types/states.py]: Maybe leave a comment: "TODO: check if `justification_bitfield` is `bytes` or `int`".

File: eth/beacon/types/validator_records.py [hunk 9]
```diff
@@ -41,12 +45,12 @@ class ValidatorRecord(rlp.Serializable):
...
 
     def __init__(self,
-                 pubkey: int,
+                 pubkey: BLSPubkey,
                  withdrawal_credentials: Hash32,
                  randao_commitment: Hash32,
-                 randao_layers: int,
-                 status: int,
-                 latest_status_change_slot: int,
+                 randao_layers: SlotNumber,
```
Comment (hwwhww) [inline eth/beacon/types/validator_records.py]: Ah, `randao_layers` is the "5 slots" thing.
```suggestion
                 randao_layers: int,
```

File: eth/beacon/types/validator_records.py [hunk 10]
```diff
@@ -67,18 +71,18 @@ def is_active(self) -> bool:
...
+                              latest_status_change_slot: SlotNumber) -> 'ValidatorRecord':
         """
         Return a new pending ``ValidatorRecord`` with the given fields.
         """
         return cls(
             pubkey=pubkey,
             withdrawal_credentials=withdrawal_credentials,
             randao_commitment=randao_commitment,
-            randao_layers=0,
+            randao_layers=SlotNumber(0),
```
Comment (hwwhww) [inline eth/beacon/types/validator_records.py]:
```suggestion
            randao_layers=0,
```

File: eth/beacon/types/proposer_slashings.py [hunk 11]
```diff
@@ -27,13 +30,14 @@ class ProposerSlashing(rlp.Serializable):
     def __init__(self,
                  proposer_index: int,
```
Comment (hwwhww) [inline eth/beacon/types/proposer_slashings.py]:
```suggestion
                 proposer_index: ValidatorIndex,
```

File: eth/beacon/helpers.py [hunk 12]
```diff
@@ -284,7 +295,7 @@ def get_block_committees_info(parent_block: 'BaseBeaconBlock',
 
 
 def get_beacon_proposer_index(state: 'BeaconState',
-                              slot: int,
+                              slot: SlotNumber,
                               epoch_length: int) -> int:
```
Comment (hwwhww) [inline eth/beacon/helpers.py]:
```suggestion
                              epoch_length: int) -> ValidatorIndex:
```

**Comment by ChihChengLiang** [comment]
Merge first and address the issues later

**Comment by cburgdorf** [comment]
>Should we use a type `ShardCount` for variable `SHARD_COUNT`, or should it share the same type with `ShardNumber`.

I would not make `SHARD_COUNT` be of type `ShardNumber` as the intend is different imho. However, I would consider just using `int` for `SHARD_COUNT`.

But that's my gut feeling response without having too much insight how it ends up in code. Like, do we have a lot of APIs accepting or returning shard counts? If yes, a `ShardCount` type might make sense, if not, just using `int` strikes me as just fine.

**Pull request merged**